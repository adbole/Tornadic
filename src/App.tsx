//Components
import Now from 'Components/Now';
import Hourly from 'Components/Hourly';
import Daily from 'Components/Daily';
import Radar from 'Components/Radar';
import Wind from 'Components/Wind';
import SunTime from 'Components/SunTime';
import Alert from 'Components/Alert';
import Pressure from 'Components/Pressure';
import { AirUV, DayValues } from 'Components/SimpleComponents';

//Contexts
import ModalContext from 'Components/Contexts/ModalContext';
import WeatherContext from 'Components/Contexts/Weather';

const App = () => (
    <>
        <WeatherContext>
            <ModalContext>
                    <Now/>
                    <Daily/>
                    <Radar/>

                    <Alert/>
                    <Hourly/>

                    <DayValues/>
                    <AirUV/>
                    <Wind/>
                    <Pressure/>
                    <SunTime/>
            </ModalContext>
        </WeatherContext>
    </>
);

export default App;
