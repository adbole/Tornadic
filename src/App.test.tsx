import { userTest } from "@test-utils";

import { render, screen } from "@testing-library/react";
import type { useUserLocation } from "Hooks";

import App from "App";


type FetchErrorHandlerProps = {
    children: React.ReactNode;
    errorRender: (hasError: boolean, retry: () => void) => React.ReactNode;
};

type WeatherContextProps = {
    latitude: number;
    longitude: number;
    children: React.ReactNode;
    skeleton: React.ReactNode;
};

const componentMocks = vi.hoisted(() => ({
    FetchErrorHandler: vi.fn(({ children }: FetchErrorHandlerProps) => (
        <div>FetchErrorHandler - {children}</div>
    )),
    WeatherContext: vi.fn(({ children }: WeatherContextProps) => (
        <div>WeatherContext - {children}</div>
    )),
}));

vi.mock("Components", async importOriginal => ({
    ...((await importOriginal()) as any),
    Alert: ({ noNotify = false }) => <div>Alert - No Notify: {noNotify.toString()}</div>,
    Background: () => <div>Background</div>,
    Daily: () => <div>Daily</div>,
    FetchErrorHandler: componentMocks.FetchErrorHandler,
    HazardLevel: ({ hazard }: { hazard: string }) => <div>HazardLevel - {hazard}</div>,
    Hourly: () => <div>Hourly</div>,
    LocationInput: () => <div>LocationInput</div>,
    Modal: ({
        children,
        isOpen,
        onClose,
    }: {
        children: React.ReactNode;
        isOpen: boolean;
        onClose: () => void;
    }) =>
        isOpen && (
            <dialog open>
                Modal
                <button onClick={onClose}>Invoke onClose</button>
                {children}
            </dialog>
        ),
    ModalContent: ({ children }: { children: React.ReactNode }) => (
        <div>ModalContent - {children}</div>
    ),
    Now: ({ displayOnly = false }) => <div>Now - Display Only: {displayOnly.toString()}</div>,
    Pressure: () => <div>Pressure</div>,
    Radar: () => <div>Radar</div>,
    Simple: ({
        icon,
        title,
        property,
    }: {
        icon: React.ReactNode;
        title: string;
        property: string;
    }) => (
        <div>
            {icon} - {title} - {property}
        </div>
    ),
    Skeleton: ({ className, size }: { className: string; size: string }) => (
        <div className={className}>Skeleton - {size}</div>
    ),
    SunTime: () => <div>SunTime</div>,
    Toast: ({
        isOpen,
        actions,
        children,
    }: {
        isOpen: boolean;
        actions: { content: string; onClick: () => void }[];
        children: React.ReactNode;
    }) =>
        isOpen && (
            <div>
                Toast
                {actions?.map(({ content, onClick }) => (
                    <button onClick={onClick} key={content}>
                        {content}
                    </button>
                ))}
                {children}
            </div>
        ),
    Wind: () => <div>Wind</div>,
}));

vi.mock("Contexts/WeatherContext", async importOriginal => ({
    ...((await importOriginal()) as any),
    default: componentMocks.WeatherContext,
}));

vi.mock("svgs", () => ({
    Spinner: () => <div>Spinner</div>,
    WifiOff: () => <div>WifiOff</div>,
}));

vi.mock("svgs/radar", async importOriginal => ({
    ...((await importOriginal()) as any),
    Cursor: () => <div>Cursor</div>,
}));

vi.mock("svgs/widget", () => ({
    Droplet: () => <>Droplet</>,
    Thermometer: () => <>Thermometer</>,
    Moisture: () => <>Moisture</>,
    Eye: () => <>Eye</>,
}));

const hookMocks = vi.hoisted(() => ({
    useOnlineOffline: vi.fn(() => true),
    useUserLocation: vi.fn(() => ({ status: "no_value" })),
}));

vi.mock("Hooks", async importOriginal => ({
    ...((await importOriginal()) as any),
    useOnlineOffline: hookMocks.useOnlineOffline,
    useUserLocation: hookMocks.useUserLocation,
}));

test("When the app is offline, a message is shown", () => {
    hookMocks.useOnlineOffline.mockReturnValueOnce(false);

    render(<App />);

    expect.soft(screen.queryByText(/Tornadic requires an internet connection/)).toBeInTheDocument();
});

describe("Location", () => {
    userTest("When there is no location, a message is shown to set it", async ({ user }) => {
        render(<App />);

        expect.soft(screen.queryByText("Cursor")).toBeInTheDocument();
        expect
            .soft(screen.queryByText(/Tornadic requires you to provide a location/))
            .toBeInTheDocument();

        await user.click(screen.getByText(/Provide Location/));

        expect.soft(screen.queryByRole("dialog")).toBeInTheDocument();
        expect.soft(screen.queryByText(/ModalContent/)).toBeInTheDocument();
        expect.soft(screen.queryByText("LocationInput")).toBeInTheDocument();

        await user.click(screen.getByText(/Invoke onClose/));

        expect.soft(screen.queryByRole("dialog")).not.toBeInTheDocument();
    });

    test.each([
        [
            "denied",
            "You have denied location access. To use your current location, please enable location access in your browser settings and refresh the page.",
        ],
        [
            "unavailable",
            "Your location could not be determined at this time. Please try again later.",
        ],
        [
            "timeout",
            "Your location could not be determined in a timely manner. Please try again later.",
        ],
        [
            "nav_not_supported",
            "Your browser does not support location services. Please use a different browser or device to use your current location.",
        ],
    ] as [ReturnType<typeof useUserLocation>["status"], string][])(
        `When the location status is %s, a message is shown`,
        (status, msg) => {
            hookMocks.useUserLocation.mockReturnValue({ status });

            render(<App />);

            expect
                .soft(screen.queryByText(/Tornadic requires you to provide a location/))
                .toBeInTheDocument();
            expect.soft(screen.getByText(msg)).toBeInTheDocument();
        }
    );

    test("When getting the location a spinner is shown with a message", () => {
        hookMocks.useUserLocation.mockReturnValueOnce({ status: "getting_current" });

        render(<App />);

        expect.soft(screen.queryByText("Spinner")).toBeInTheDocument();
        expect.soft(screen.queryByText("Getting Your Location")).toBeInTheDocument();
    });

    //Normally, the WeatherContext would enter is loading state using a skeleton, however this is ensuring
    //that App is using all the right components when its checks have passed
    test("When the location is available, the app renders", () => {
        hookMocks.useUserLocation.mockReturnValueOnce({
            latitude: 0,
            longitude: 0,
            status: "OK",
        } as any);

        render(<App />);

        expect.soft(screen.queryByText(/FetchErrorHandler/)).toBeInTheDocument();
        expect.soft(screen.queryByText(/WeatherContext/)).toBeInTheDocument();
        expect.soft(screen.queryByText("Now - Display Only: false")).toBeInTheDocument();
        expect.soft(screen.queryByText("Alert - No Notify: false")).toBeInTheDocument();
        expect.soft(screen.queryByText("Hourly")).toBeInTheDocument();
        expect.soft(screen.queryByText("Daily")).toBeInTheDocument();
        expect.soft(screen.queryByText("Radar")).toBeInTheDocument();

        expect
            .soft(screen.queryByText("Droplet - Precipitation - precipitation"))
            .toBeInTheDocument();
        expect.soft(screen.queryByText("Thermometer - Dewpoint - dewpoint_2m")).toBeInTheDocument();
        expect
            .soft(screen.queryByText("Moisture - Humidity - relativehumidity_2m"))
            .toBeInTheDocument();

        expect.soft(screen.queryByText("HazardLevel - us_aqi")).toBeInTheDocument();
        expect.soft(screen.queryByText("HazardLevel - uv_index")).toBeInTheDocument();

        expect.soft(screen.queryByText("Wind")).toBeInTheDocument();
        expect.soft(screen.queryByText("Pressure")).toBeInTheDocument();
        expect.soft(screen.queryByText("SunTime")).toBeInTheDocument();

        expect.soft(screen.queryByText(/Could not get weather data/)).not.toBeInTheDocument();

        expect.soft(componentMocks.WeatherContext).toHaveBeenCalledWith(
            expect.objectContaining({
                latitude: 0,
                longitude: 0,
            }),
            expect.anything()
        );
    });
});

userTest("FetchErrorHandler's errorRender renders a toast and correctly provides retry", async ({ user }) => {
    const retry = vi.fn();

    componentMocks.FetchErrorHandler.mockImplementationOnce(
        ({ children, errorRender }: FetchErrorHandlerProps) => (
            <div>
                FetchErrorHandler -{children}
                {errorRender(true, retry)}
            </div>
        )
    );

    hookMocks.useUserLocation.mockReturnValueOnce({
        latitude: 0,
        longitude: 0,
        status: "OK",
    } as any);

    render(<App />);

    expect.soft(screen.queryByText(/Could not get weather data/)).toBeInTheDocument();

    await user.click(screen.getByText("Try Again Now"));

    expect.soft(retry).toHaveBeenCalledOnce();
});

test("The weather context is provided with a skeleton", () => {
    componentMocks.WeatherContext.mockImplementationOnce(({ skeleton }: WeatherContextProps) => (
        <div>WeatherContext - {skeleton}</div>
    ));

    hookMocks.useUserLocation.mockReturnValueOnce({
        latitude: 0,
        longitude: 0,
        status: "OK",
    } as any);

    const { container } = render(<App />);

    const skeletons = screen.queryAllByText(/Skeleton/);

    expect(skeletons).toHaveLength(13);
    expect(container).toMatchSnapshot();
});
