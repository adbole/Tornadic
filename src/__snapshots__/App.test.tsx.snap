// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`The weather context is provided with a skeleton 1`] = `
.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 10px;
  font-size: 1.5rem;
  padding: 60px 0px;
}

@media screen and (max-width: 1100px) {
  .emotion-0 {
    grid-column: 1/-1;
  }
}

@media screen and (min-width: 1100px) {
  #root>.emotion-0 {
    grid-area: n;
  }
}

@media screen and (min-width: 800px) {
  .emotion-1 {
    grid-column: span 4;
  }

  .e1r27hfx2+.emotion-1 {
    grid-column: span 2;
  }
}

@media screen and (min-width: 1400px) {
  .emotion-1 {
    grid-column: span 6;
  }

  .e1r27hfx2+.emotion-1 {
    grid-column: span 4;
  }
}

@media screen and (min-width: 1100px) {
  #root>.emotion-2 {
    grid-area: d;
  }
}

.emotion-3 {
  -webkit-backdrop-filter: none;
  backdrop-filter: none;
}

.emotion-3:before {
  content: '';
  position: absolute;
  inset: 0;
  z-index: -1;
  -webkit-backdrop-filter: blur(10px);
  backdrop-filter: blur(10px);
  box-shadow: inset 0 0 0 200px rgba(255, 255, 255, 0.01);
}

.emotion-3 .leaflet-center {
  left: 0;
  right: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: auto;
  gap: 10px;
  padding: 10px;
  max-width: 900px;
}

.emotion-3 .leaflet-control {
  background-color: rgba(0, 0, 0, 0.75);
  -webkit-backdrop-filter: blur(10px);
  backdrop-filter: blur(10px);
  color: white;
  border-radius: var(--border-radius);
  overflow: hidden;
}

.emotion-3 .leaflet-custom-control,
.emotion-3 .leaflet-control-toggle {
  background-color: rgba(0, 0, 0, 0.75);
  -webkit-backdrop-filter: blur(10px);
  backdrop-filter: blur(10px);
  color: white;
  color: white!important;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 2rem;
  min-height: 2rem;
  border: none;
  --svg-size: 1.25rem;
}

.emotion-3 .leaflet-overlay-pane svg {
  width: unset;
}

@media screen and (min-width: 1100px) {
  .emotion-3 {
    grid-area: r;
  }
}

.emotion-4 {
  display: grid;
  gap: inherit;
}

<div>
  <div>
    FetchErrorHandler - 
    <div>
      WeatherContext - 
      <div
        class="emotion-0"
      >
        Skeleton - 
        widget-large
      </div>
      <div
        class="emotion-1"
      >
        Skeleton - 
        widget-wide
      </div>
      <div
        class="emotion-2"
      >
        Skeleton - 
        widget-large
      </div>
      <div
        class="emotion-3"
      >
        Skeleton - 
        widget-large
      </div>
      <div
        class="emotion-4 emotion-5"
      >
        <div>
          Skeleton - 
        </div>
        <div>
          Skeleton - 
        </div>
      </div>
      <div
        class="emotion-4 emotion-5"
      >
        <div>
          Skeleton - 
        </div>
        <div>
          Skeleton - 
        </div>
      </div>
      <div>
        Skeleton - 
      </div>
      <div>
        Skeleton - 
      </div>
      <div>
        Skeleton - 
        widget-wide
      </div>
      <div>
        Skeleton - 
      </div>
      <div>
        Skeleton - 
      </div>
    </div>
  </div>
</div>
`;
